### PR TITLE
Edit account

### PR DESCRIPTION
**Pull request by rachek-o**

### CODE REVIEW

File: src/main/java/com/opencart/datamodel/RegisterModel.java [hunk 1]
```diff
@@ -1,54 +1,48 @@
...
-        public RegisterModel build() {
+        public RegisterModel build(){
             return RegisterModel.this;
         }
     }
 }
+
+
+
+
```
Comment (adomanich) [inline src/main/java/com/opencart/datamodel/RegisterModel.java:54]: Please, remove extra spaces

File: src/main/java/com/opencart/enums/URLs.java [hunk 2]
```diff
@@ -11,6 +10,7 @@
     }
 
     public String getValue() {
+
```
Comment (adomanich) [inline src/main/java/com/opencart/enums/URLs.java]: Please, remove extra spaces

File: src/main/java/com/opencart/pages/EditAccountPage.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    private WebElement errorMessage;
+
+    @FindBy(xpath = "//div[contains(@class, 'text-danger') and text()='Last Name must be between 1 and 32 characters!']")
+    private WebElement warningMessage;
+
+    @FindBy(css  = "#content > form > fieldset > div:nth-child(4) > div > div")
+    private WebElement cautionMessage;
+
+    @FindBy(css = "#content > form > fieldset > div:nth-child(5) > div > div")
+    private WebElement dangerMessage;
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/EditAccountPage.java]: Try to rewrite it without using "magic numbers" in your locators

File: src/main/java/com/opencart/pages/EditAccountPage.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        return warningMessage;
+    }
+
+    public WebElement getCautionMessage() {
+        return cautionMessage;
+    }
+    public WebElement getDangerMessage() {
+        return dangerMessage;
+    }
+
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/EditAccountPage.java:80]: Please, remove extra spaces

File: src/main/java/com/opencart/pages/HeaderPage.java [hunk 5]
```diff
@@ -12,6 +12,12 @@
     @FindBy(xpath = ".//*[contains(@href, 'register')]")
     private WebElement registerButton;
 
+    @FindBy(xpath = ".//*[contains(@href, 'login')]")
+    private WebElement loginPageButton;
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/HeaderPage.java]: Please, rename it to loginButton

File: src/main/java/com/opencart/pages/HeaderPage.java [hunk 6]
```diff
@@ -20,4 +26,9 @@ public WebElement getMyAccountButton() {
     public WebElement getRegisterButton() {
         return registerButton;
     }
+
+    public WebElement getLoginPageButton() {
+        return loginPageButton;
+    }
+
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/HeaderPage.java]: Please, remove extra space

File: src/main/java/com/opencart/pages/LoginPage.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    public WebElement getEmailInput()  {
+        return emailInput;
+    }
+
+    public WebElement getPasswordInput() {
+        return passwordInput;
+    }
+
+    public WebElement getLoginButton() throws InterruptedException {
+        Thread.sleep(5000);
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/LoginPage.java]: Using Thread.sleep() is a bad practice, try to replace it with explicit wait

File: src/main/java/com/opencart/pages/LoginPage.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        return loginButton;
+    }
+
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/LoginPage.java:27]: Please, remove extra space

File: src/main/java/com/opencart/pages/MyAccountPage.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class MyAccountPage extends HeaderPage {
+    @FindBy(xpath = ".//*[contains(@href, 'edit')]")
+    private WebElement editAccountButton;
+
+    public WebElement getEditAccountButton() throws InterruptedException {
+        Thread.sleep(5000);
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/MyAccountPage.java]: Using Thread.sleep() is a bad practice, try to replace it with explicit wait

File: src/main/java/com/opencart/pages/MyAccountPage.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+    public WebElement getAccountTab() throws InterruptedException {
+        Thread.sleep(5000);
+        return accountTab;
+    }
+
+
+    }
+
+
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/MyAccountPage.java]: Please, remove extra spaces

File: src/main/java/com/opencart/pages/MyAccountPage.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    private WebElement successMessage;
+
+    public WebElement getSuccessMessage() {
+        return successMessage;
+    }
+    @FindBy(xpath = ".//*[contains(@href, 'account')]")
+    private WebElement accountTab;
+
+    public WebElement getAccountTab() throws InterruptedException {
+        Thread.sleep(5000);
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/MyAccountPage.java]: Using Thread.sleep() is a bad practice, try to replace it with explicit wait

File: src/main/java/com/opencart/pages/MyAccountPage.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+public class MyAccountPage extends HeaderPage {
+    @FindBy(xpath = ".//*[contains(@href, 'edit')]")
+    private WebElement editAccountButton;
+
+    public WebElement getEditAccountButton() throws InterruptedException {
+        Thread.sleep(5000);
+        return editAccountButton;
+    }
+
+    @FindBy(xpath = ".//div[@class='alert alert-success alert-dismissible']")
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/MyAccountPage.java:15]: you can use xpath - contains() function and define just a part of the full class name in your locator

File: src/main/java/com/opencart/pages/SuccessRegisterPage.java [hunk 13]
```diff
@@ -4,11 +4,11 @@
...
 
 public class SuccessRegisterPage extends BasePage {
-
     @FindBy(xpath = ".//*[@id='content']/h1")
     private WebElement successTitle;
 
     public WebElement getSuccessTitle() {
         return successTitle;
     }
+
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/SuccessRegisterPage.java:15]: Please, remove extra space

File: src/main/java/com/opencart/repository/EditAccountModelRepository.java [hunk 14]
```diff
@@ -0,0 +1,19 @@
+package com.opencart.repository;
+
+import com.opencart.datamodel.EditAccountModel;
+import org.apache.commons.lang3.RandomStringUtils;
+
+public class EditAccountModelRepository {
+        private EditAccountModelRepository(){
+
```
Comment (adomanich) [inline src/main/java/com/opencart/repository/EditAccountModelRepository.java]: Please, remove extra space

File: src/main/java/com/opencart/repository/EditAccountModelRepository.java [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+        }
+        public static EditAccountModel getEditAccountModel(){
```
Comment (adomanich) [inline src/main/java/com/opencart/repository/EditAccountModelRepository.java]: Please, add space here

File: src/main/java/com/opencart/repository/LoginModelRepository.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
+package com.opencart.repository;
+
+import com.opencart.datamodel.LoginModel;
+
+public class LoginModelRepository {
+    private LoginModelRepository(){
+
```
Comment (adomanich) [inline src/main/java/com/opencart/repository/LoginModelRepository.java]: Please, remove extra space

File: src/main/java/com/opencart/repository/LoginModelRepository.java [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+    }
+    public static LoginModel getLoginModel(){
```
Comment (adomanich) [inline src/main/java/com/opencart/repository/LoginModelRepository.java]: Please, add space here

File: src/main/java/com/opencart/steps/EditAccountPageBL.java [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+        EditAccountModel editAccountModel = EditAccountModelRepository.getEditAccountModel();
+        inputFirstName(editAccountModel.getFirstName());
+        inputLastName(editAccountModel.getLastName());
+        inputEmail(editAccountModel.getEmail());
+        inputTelephone(editAccountModel.getTelephone());
+
+        clickOnBackButton();
+
+        myAccountPage = new MyAccountPage();
+        return this;
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/EditAccountPageBL.java:64]: I believe it should return new MyAccountPageBL()

File: src/main/java/com/opencart/steps/EditAccountPageBL.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+        String expectedMessage = "E-Mail Address does not appear to be valid!";
+        Assert.assertEquals(editAccountPage.getCautionMessage().getText(), expectedMessage);
+    }
+    public void lackOfTelephoneCharacters(){
+        String expectedMessage = "Telephone must be between 3 and 32 characters!";
+        Assert.assertEquals(editAccountPage.getDangerMessage().getText(), expectedMessage);
+    }
+    public void redirectToPreviousPage() throws InterruptedException {
+        Assert.assertTrue(Boolean.parseBoolean(String.valueOf(myAccountPage.getAccountTab().isDisplayed())));
+    }
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/EditAccountPageBL.java]: All business logic  methods which are responsible for verifying something should start with "verify" or "check" prefix

File: src/main/java/com/opencart/steps/EditAccountPageBL.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    public void incorrectEmail(){
+        String expectedMessage = "E-Mail Address does not appear to be valid!";
+        Assert.assertEquals(editAccountPage.getCautionMessage().getText(), expectedMessage);
+    }
+    public void lackOfTelephoneCharacters(){
+        String expectedMessage = "Telephone must be between 3 and 32 characters!";
+        Assert.assertEquals(editAccountPage.getDangerMessage().getText(), expectedMessage);
+    }
+    public void redirectToPreviousPage() throws InterruptedException {
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/EditAccountPageBL.java]: Why do you throw InterruptedException here?

File: src/main/java/com/opencart/steps/HeaderPageBL.java [hunk 21]
```diff
@@ -19,4 +22,12 @@ public RegisterPageBL clickOnRegisterButton() {
...
     }
+    public LoginPageBL clickOnLoginPageButton() {
+        headerPage.getLoginPageButton().click();
+        return new LoginPageBL();}
 }
+
+
+
+
+
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/HeaderPageBL.java]: Please, remove extra spaces

File: src/main/java/com/opencart/steps/HeaderPageBL.java [hunk 22]
```diff
@@ -1,12 +1,15 @@
...
 import com.opencart.pages.HeaderPage;
+import com.opencart.pages.MyAccountPage;
+import org.openqa.selenium.WebElement;
 
 public class HeaderPageBL {
 
     private HeaderPage headerPage;
 
     public HeaderPageBL() {
+
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/HeaderPageBL.java]: Please, remove extra space

File: src/main/java/com/opencart/steps/LoginPageBL.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+    public LoginPageBL() {
+        loginPage = new LoginPage();
+    }
+
+    public LoginPageBL loginPerson() {
+        LoginModel loginModel = LoginModelRepository.getLoginModel();
+        inputEmail(loginModel.getEmail());
+        inputPassword(loginModel.getPassword());
+
+        myAccountPage = new MyAccountPage();
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/LoginPageBL.java:19]: It should be created in constructor within MyAccountPageBL class

File: src/main/java/com/opencart/steps/MyAccountPageBL.java [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+
+    public MyAccountPageBL() {
+
+        myAccountPage = new MyAccountPage();
+    }
+    public EditAccountPageBL clickOnEditAccountButton() throws InterruptedException {
+        myAccountPage.getEditAccountButton().click();
+        return new EditAccountPageBL();
+    }
+
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/MyAccountPageBL.java:17]: Please, remove extra space

File: src/main/java/com/opencart/steps/MyAccountPageBL.java [hunk 25]
```diff
@@ -0,0 +1,17 @@
+package com.opencart.steps;
+
+import com.opencart.pages.MyAccountPage;
+
+public class MyAccountPageBL {
+    private MyAccountPage myAccountPage;
+
+    public MyAccountPageBL() {
+
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/MyAccountPageBL.java:9]: Please, remove extra space

File: src/test/java/EditAccountTest.java [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+                .cancelEditInformation();
+        editAccountPageBL.redirectToPreviousPage();
+    }
+
+    @Test
+    public void updateUserParameters() {
+        EditAccountPageBL editAccountPageBL = new EditAccountPageBL()
+                .editInformation();
+        editAccountPageBL.verifyUpdates();
+    }
```
Comment (adomanich) [inline src/test/java/EditAccountTest.java]: You can perform all negative validations within one test and also, in addition, create a positive one with valid edit data

**Review by adomanich — CHANGES_REQUESTED** [review]

**Review by adomanich — APPROVED** [review]

**Review by Alex7mob1 — APPROVED** [review]

**Pull request merged**